control-net-qrcode/model/model.py

```diff
@@ -0,0 +1,103 @@
+import torch
+import base64
+import qrcode
+from PIL import Image
+from io import BytesIO
+from typing import Dict
+
+
+from diffusers import (
+    StableDiffusionControlNetPipeline,
+    ControlNetModel,
+    DDIMScheduler,
+    DPMSolverMultistepScheduler,
+    DEISMultistepScheduler,
+    HeunDiscreteScheduler,
+    EulerDiscreteScheduler,
+    EulerAncestralDiscreteScheduler
+)
+
+
+class Model:
+    def __init__(self, **kwargs):
+        self.model = None
+
+    def create_code(self, content: str):
+        qr = qrcode.QRCode(
+            version=1,
+            error_correction=qrcode.constants.ERROR_CORRECT_H,
+            box_size=16,
+            border=0,
+        )
+        qr.add_data(content)
+        qr.make(fit=True)
+        img = qr.make_image(fill_color="black", back_color="white")
+
+        offset_min = 8 * 16
+        w, h = img.size
+        w = (w + 255 + offset_min) // 256 * 256
+        h = (h + 255 + offset_min) // 256 * 256
+        bg = Image.new('L', (w, h), 128)
+
+        coords = ((w - img.size[0]) // 2 // 16 * 16,
+                  (h - img.size[1]) // 2 // 16 * 16)
+        bg.paste(img, coords)
+        return bg
+
+    def pil_to_b64(self, pil_img):
+        buffered = BytesIO()
+        pil_img.save(buffered, format="PNG")
+        img_str = base64.b64encode(buffered.getvalue()).decode("utf-8")
+        return img_str
+
+    def load(self):
+        controlnet = ControlNetModel.from_pretrained(
+            "monster-labs/control_v1p_sd15_qrcode_monster", torch_dtype=torch.float16
+        )
+
+        self.model = StableDiffusionControlNetPipeline.from_pretrained(
+            "runwayml/stable-diffusion-v1-5",
+            controlnet=controlnet,
+            safety_checker=None,
+            torch_dtype=torch.float16,
+        ).to("cuda")
+        self.model.enable_xformers_memory_efficient_attention()
+
+    def predict(self, request: Dict) -> Dict:
+        qr_code_content = request.get("qr_code_content")
```

Comment:
the content can be either:
- a url to turn into a qr code
- OR more generally, some kind of mask so users can make images like this: https://twitter.com/MrUgleh/status/1702041188482658758?s=20